### PR TITLE
Track JVM gc time in histogram

### LINKED ISSUE #2902: Track JVM gc time
Related to #2900.

Its valuable to track the time spent in garbage collection, how many garbage collection events took place, and the type of garbage collection events that took place (different garbage collector implementations have various phases which happen concurrently or in parallel). 

I proposed [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6362) that we use a histogram to collect this data:

| Name | Description | Unit | Unit UCUM | Instrument Type | Value Type | Attribute Key | Attribute Values  |
|---|---|---|---|---|---|---|---|
| process.runtime.jvm.gc.time | Time spent performing JVM garbage collection actions | Milliseconds | `ms`                                      | Histogram | Int64 | gc | Name of garbage collector |
| | | | | | | action | The collection action performed |

Recording gc event times as a histogram would enable the following analysis:
* Total time spent on garbage collection.
* Count of garbage collection events.
* Min, max, average, percentiles of garbage collection time events.
* The attributes allow you to characterize gc time by the types of events. With a little prior knowledge you can analyze all gc spent concurrently vs stop the world.

### PR DESCRIPTION
**Pull request by jack-berg**
Fixes #2902.

cc @open-telemetry/java-instrumentation-maintainers, @open-telemetry/java-maintainers

### CODE REVIEW

File: specification/metrics/semantic_conventions/runtime-environment-metrics.md [hunk 1]
```diff
@@ -61,28 +61,44 @@ consider, for example pthreads vs green thread implementations.
...
 **[2]**: These utilizations are not defined as being for the specific interval since last measurement (unlike `system.cpu.utilization`).
+
 **[3]**: Pool names are generally obtained via [BufferPoolMXBean#getName()](https://docs.oracle.com/en/java/javase/11/docs/api/java.management/java/lang/management/BufferPoolMXBean.html#getName()).
+
+**[4]**: Garbage collector name and action are generally obtained via [GarbageCollectionNotificationInfo#getGcName()](https://docs.oracle.com/en/java/javase/11/docs/api/jdk.management/com/sun/management/GarbageCollectionNotificationInfo.html#getGcName()) and [GarbageCollectionNotificationInfo#getGcAction()](https://docs.oracle.com/en/java/javase/11/docs/api/jdk.management/com/sun/management/GarbageCollectionNotificationInfo.html#getGcAction()) respectively. The following is an incomplete list of names and actions for several popular garbage collectors:
+
+* G1 Garbage Collector
+  * Names: `G1 Young Generation`, `G1 Old Generation`
+  * Actions: `end of minor GC`, `end of major GC`
+* Senandoah Garbage Collector
```
Comment (trask) [inline specification/metrics/semantic_conventions/runtime-environment-metrics.md:99]:
```suggestion
* Shenandoah Garbage Collector
```

**Review by trask — APPROVED** [review]

**Review by bogdandrutu — COMMENTED** [review]
Why do you think that a default of histogram is better?

File: specification/metrics/semantic_conventions/runtime-environment-metrics.md [hunk 2]
```diff
@@ -61,28 +61,44 @@ consider, for example pthreads vs green thread implementations.
...
+| process.runtime.jvm.buffer.usage           | Measure of memory used by buffers                        | Bytes        | `By`                                      | UpDownCounter                                     | Int64      | pool          | Name of pool[3]       |
+| process.runtime.jvm.buffer.limit           | Measure of total memory capacity of buffers              | Bytes        | `By`                                      | UpDownCounter                                     | Int64      | pool          | Name of pool[3]       |
+| process.runtime.jvm.buffer.count           | Number of buffers in the pool                            | buffers      | `{buffers}`                               | UpDownCounter                                     | Int64      | pool          | Name of pool[3]       |
+| process.runtime.jvm.gc.time                | Time spent performing JVM garbage collection actions     | Milliseconds | `Ms`                                      | Histogram                                         | Int64      | gc            | Name of gc[4]         |
```
Comment (mateuszrzeszutek) [inline specification/metrics/semantic_conventions/runtime-environment-metrics.md:84]: Nit
```suggestion
| process.runtime.jvm.gc.time                | Time spent performing JVM garbage collection actions     | Milliseconds | `ms`                                      | Histogram                                         | Int64      | gc            | Name of gc[4]         |
```

**Review by mateuszrzeszutek — APPROVED** [review]

**Comment by jack-berg** [comment]
> Why do you think that a default of histogram is better?

@bogdandrutu not sure I understand. Better than what?

In general, histogram is a good fit for this because it includes a lot of useful data for analysis: 
- You can use the sum to analyze the total time spent in gc
- Can look at the count to analyze the number of gc events
- Can look at the max to analyze the maximum time spent in any single gc event (i.e. what was my longest "stop the world" gc event?)
- Can look at the distribution to analyze the distribution of time of gc events (i.e. how long do typical gc events take? how long do outlier gc events take?)

**Review by breedx-splk — APPROVED** [review]

**Comment by trask** [comment]
@bogdandrutu just wanted to make sure you don't have any significant concerns before we request merging, thx!

**Review by jmacd — APPROVED** [review]

File: specification/metrics/semantic_conventions/runtime-environment-metrics.md [hunk 3]
```diff
@@ -61,30 +61,46 @@ consider, for example pthreads vs green thread implementations.
...
+| process.runtime.jvm.buffer.limit               | Measure of total memory capacity of buffers                                                     | Bytes        | `By`                                      | UpDownCounter                                     | Int64      | pool          | Name of pool[3]       |
+| process.runtime.jvm.buffer.count               | Number of buffers in the pool                                                                   | buffers      | `{buffers}`                               | UpDownCounter                                     | Int64      | pool          | Name of pool[3]       |
+| process.runtime.jvm.gc.time                    | Time spent performing JVM garbage collection actions                                            | Milliseconds | `ms`                                      | Histogram                                         | Int64      | gc            | Name of gc[4]         |
```
Comment (jmacd) [inline specification/metrics/semantic_conventions/runtime-environment-metrics.md:86]: As an aside, I wonder if the paragraph describing `.time` here https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#instrument-naming should be extended to cover Histogram measurements of time?

Comment (trask) [reply specification/metrics/semantic_conventions/runtime-environment-metrics.md:86]: maybe histogram measurements of time could be `.duration` (similar to `http.server.duration`)?

**Comment by jack-berg** [comment]
FYI I've updated the metric name to `process.runtime.jvm.gc.duration`, prompted by [this comment](https://github.com/open-telemetry/opentelemetry-specification/pull/2903#discussion_r1018305894). I quite like @trask's suggestion to use the duration suffix instead of time. Makes it clear that this metric measures the distribution of individual gc event durations, rather than the total time spent in in garbage collection. 

Please take another look approvers @breed-splk, @mateuszrzeszutek @jmacd @trask

**Review by mateuszrzeszutek — APPROVED** [review]

**Review by trask — APPROVED** [review]

**Comment by trask** [comment]
@bogdandrutu @jmacd @reyang @jsuereth - the Java folks would like to merge this if there's no further concerns

**Review by reyang — APPROVED** [review]

**Comment by reyang** [comment]
There is a wider discussion brought up by @jsuereth this morning regarding whether we should have different way of exposing metrics for the same concept or not. We should continue that discussion in the semantic convention workgroup, no need to block this PR since it has strong support from Java folks.

**Pull request merged**